components/script/dom/window.rs

```diff
@@ -684,37 +739,72 @@ impl WindowMethods for Window {
 
     // https://html.spec.whatwg.org/multipage/#dom-document-2
     fn Document(&self) -> DomRoot<Document> {
-        self.document
-            .get()
-            .expect("Document accessed before initialization.")
+        let document = match &*self.current_state.borrow() {
+            WindowState::Zombie(document) => document.get(),
+            WindowState::Alive(data) => data.document.get(),
+        };
+        document.expect("Document accessed before initialization.")
     }
 
     // https://html.spec.whatwg.org/multipage/#dom-history
     fn History(&self) -> DomRoot<History> {
-        self.history.or_init(|| History::new(self))
+        match &*self.current_state.borrow() {
+            // In the Zombie case, create an object for script, but don't store it.
+            // Usage of the object should throw security errors,
+            // since there is no associated document or browsing-context.
+            WindowState::Zombie(_) => History::new(self),
+            WindowState::Alive(data) => data.history.or_init(|| History::new(self)),
+        }
     }
 
     // https://html.spec.whatwg.org/multipage/#dom-window-customelements
     fn CustomElements(&self) -> DomRoot<CustomElementRegistry> {
-        self.custom_element_registry
-            .or_init(|| CustomElementRegistry::new(self))
+        match &*self.current_state.borrow() {
+            // In the Zombie case, create an object for script, but don't store it.
+            // Usage of the object should throw security errors,
+            // since there is no associated document or browsing-context.
+            WindowState::Zombie(_) => CustomElementRegistry::new(self),
```

Comment:
The comment here looks wrong, since it seems the registry does not have a concept of throwing errors when used after the BC as been discarded. However it could still be reachable from script even in the zombie state. 

The idea is that if we still make the interface available to script, but don't store it, we can be sure the Rust part isn't keeping the window alive for longer than it would be used from script, including via potential references to the window from these various interfaces.